File: src/main/java/com/indra87g/commands/SetBlockCommand.java
package com.indra87g.commands;

import cn.nukkit.Player;
import cn.nukkit.block.Block;
import cn.nukkit.level.Level;
import cn.nukkit.math.Vector3;
import com.indra87g.utils.MessageHandler;

public class SetBlockCommand extends BaseCommand {

    public SetBlockCommand() {
        super(
            "setblock",
            "Place a block at specific location",
            "/setblock <x> <y> <z> <block_id>",
            "waffle.setblock"
        );
    }

    @Override
    protected boolean validateArgs(String[] args, Player player) {
        if (args.length != 4) {
            MessageHandler.sendMessage(player, "setblock_invalid_arg_count");
            MessageHandler.sendMessage(player, "invalid_usage", "{usage}", this.getUsage());
            return false;
        }

        try {
            Integer.parseInt(args[0]);
            int y = Integer.parseInt(args[1]);
            Integer.parseInt(args[2]);
            Integer.parseInt(args[3]);

            if (y < 0 || y > 255) {
                MessageHandler.sendMessage(player, "setblock_invalid_y");
                return false;
            }
        } catch (NumberFormatException e) {
            MessageHandler.sendMessage(player, "setblock_invalid_arg_type");
            MessageHandler.sendMessage(player, "invalid_usage", "{usage}", this.getUsage());
            return false;
        }
        return true;
    }

    @Override
    protected boolean handleCommand(Player player, String[] args) {
        int x = Integer.parseInt(args[0]);
        int y = Integer.parseInt(args[1]);
        int z = Integer.parseInt(args[2]);
        int blockId = Integer.parseInt(args[3]);

        Level level = player.getLevel();
        Vector3 position = new Vector3(x, y, z);
        Block block = Block.get(blockId);

        if (block.getId() == Block.AIR && blockId != 0) {
            MessageHandler.sendMessage(player, "setblock_invalid_block_id", "{id}", String.valueOf(blockId));
            return false;
        }

        boolean success = level.setBlock(position, block, true, true);

        if (success) {
            MessageHandler.sendMessage(player, "setblock_success",
                "{block}", block.getName(),
                "{x}", String.valueOf(x),
                "{y}", String.valueOf(y),
                "{z}", String.valueOf(z)
            );
            player.getServer().getLogger().info(
                player.getName() + " placed " + block.getName() +
                " at (" + x + ", " + y + ", " + z + ") in world " + level.getName()
            );
        } else {
            MessageHandler.sendMessage(player, "setblock_failure");
        }
        return true;
    }
}

File: src/main/kotlin/com/indra87g/commands/SetBlockCommand.kt
package com.indra87g.commands

import cn.nukkit.Player
import cn.nukkit.block.Block
import cn.nukkit.math.Vector3
import com.indra87g.utils.MessageHandler

class SetBlockCommand : BaseCommand(
    "setblock",
    "Place a block at specific location",
    "/setblock <x> <y> <z> <block_id>",
    "waffle.setblock"
) {

    override fun validateArgs(args: Array<String>, player: Player): Boolean {
        if (args.size != 4) {
            MessageHandler.sendMessage(player, "setblock_invalid_arg_count")
            MessageHandler.sendMessage(player, "invalid_usage", "{usage}", usageMessage)
            return false
        }

        val y = args[1].toIntOrNull()
        if (args[0].toIntOrNull() == null || y == null || args[2].toIntOrNull() == null || args[3].toIntOrNull() == null) {
            MessageHandler.sendMessage(player, "setblock_invalid_arg_type")
            MessageHandler.sendMessage(player, "invalid_usage", "{usage}", usageMessage)
            return false
        }

        if (y !in 0..255) {
            MessageHandler.sendMessage(player, "setblock_invalid_y")
            return false
        }
        return true
    }

    override fun handleCommand(player: Player, args: Array<String>): Boolean {
        val x = args[0].toInt()
        val y = args[1].toInt()
        val z = args[2].toInt()
        val blockId = args[3].toInt()

        val position = Vector3(x.toDouble(), y.toDouble(), z.toDouble())
        val block = Block.get(blockId)

        if (block.id == Block.AIR && blockId != 0) {
            MessageHandler.sendMessage(player, "setblock_invalid_block_id", "{id}", blockId.toString())
            return false
        }

        val success = player.level.setBlock(position, block, true, true)

        if (success) {
            MessageHandler.sendMessage(
                player, "setblock_success",
                "{block}", block.name,
                "{x}", x.toString(),
                "{y}", y.toString(),
                "{z}", z.toString()
            )
            player.server.logger.info(
                "${player.name} placed ${block.name} at ($x, $y, $z) in world ${player.level.name}"
            )
        } else {
            MessageHandler.sendMessage(player, "setblock_failure")
        }
        return true
    }
}
